stingray/base.py

```diff
@@ -61,36 +159,206 @@ def array_attrs(self) -> list[str]:
         ]
+        if self.not_array_attr is not None and len(self.not_array_attr) >= 1:
+            all_meta_attrs += self.not_array_attr
+        return all_meta_attrs
+
+    def dict(self) -> dict:
+        """Return a dictionary representation of the object."""
+        from collections import OrderedDict
+
+        main_attr = self.main_array_attr
+        meta_attrs = self.meta_attrs()
+        array_attrs = self.array_attrs()
+        internal_array_attrs = self.internal_array_attrs()
+
+        results = OrderedDict()
+        results[main_attr] = getattr(self, main_attr)
+
+        for attr in internal_array_attrs:
+            if isinstance(getattr(self.__class__, attr.lstrip("_"), None), property):
+                attr = attr.lstrip("_")
+            results[attr] = getattr(self, attr)
+
+        for attr in array_attrs:
+            results[attr] = getattr(self, attr)
+
+        for attr in meta_attrs:
+            results[attr] = getattr(self, attr)
+
+        return results
+
+    def pretty_print(self, func_to_apply=None, attrs_to_apply=[], attrs_to_discard=[]) -> str:
+        """Return a pretty-printed string representation of the object.
+
+        This is useful for debugging, and for interactive use.
+
+        Optional parameters
+        -------------------
+        func_to_apply : function
+            A function that modifies the attributes listed in ``attrs_to_apply``.
+            It must return the modified attributes and a label to be printed.
+            If ``None``, no function is applied.
+        attrs_to_apply : list of str
+            Attributes to be modified by ``func_to_apply``.
+        attrs_to_discard : list of str
+            Attributes to be discarded from the output.
+        """
+        print(self.__class__.__name__)
+        print("_" * len(self.__class__.__name__))
+        items = self.dict()
+        results = ""
+        np.set_printoptions(threshold=3, edgeitems=1)
+        for attr in items.keys():
+            if attr in attrs_to_discard:
+                continue
+            value = items[attr]
+            label = f"{attr:<15}: {items[attr]}"
+
+            if isinstance(value, Iterable) and not isinstance(value, str):
+                size = np.shape(value)
+                if len(size) == 1:
+                    label += f" (size {size[0]})"
+                else:
+                    label += f" (shape {size})"
+
+            if func_to_apply is not None and attr in attrs_to_apply:
+                new_value, new_label = func_to_apply(items[attr])
+                label += f"\n{attr + ' (' +new_label + ')':<15}: {new_value}"
+
+            results += label + "\n"
+        return results
+
+    def __repr__(self) -> str:
+        """Return a string representation of the object."""
+        return self.pretty_print()
+
+    def __str__(self) -> str:
+        """Return a string representation of the object."""
+        return self.__repr__()
```

Comment:
They are not exactly the same. `pretty_print` can be customized with the input values. `__str__` is what is called when the user `print`s the data, and I figured it would be good to have an idea of the data contained in the object. 
For `__repr__`, I actually re-read the docs. I thought it was just a duplicate of `__str__`, but it's probably better to leave the default python does (`<StingrayObject> at blablahaddress`)